realm-deactivation: Send email to owners when deactivating organization.

laurynmm opened the pull request:
Creates a new `realm_deactivated` email that can be sent as part of the realm deactivation process.

Fixes #24685.

**Notes**:
- Adds a new boolean flag, `email_owners`, to `do_deactivate_realm`. This flag is set to `False` when `do_deactivate_realm` is used for realm exports or changing a realm's subdomain, so that the active organization owners are not emailed in those cases.

- This flag is optional for the `deactivate_realm` management command, but as there is no active user passed in that case, then the email is sent without a reference to who deactivated the realm.

- It is passed as `True` for the support analytics view, but the email that is generated does not include information about the support user who completed the request for organization deactivation.

- When an active organization owner deactivates the organization, then the flag is `True` and an email is sent to them as well as any other active organization owners, with a slight variation in the email text for those two cases.

- Adds specific tests for when `email_owners` is passed as `True`. All existing tests for other functionality of `do_deactivate_user` pass the flag as `False`.

**Questions**:
- The date format in the email is currently `year-month-day` and isn't adjusted for the realm's default langauge. We could reorder the date parts (e.g. `day-month-year`) or create a text based string, like we do for login notification emails. But we should probably think about internationalization here. While the date string for the login notification emails is good for English, it's a bit odd in other languages (see screenshot below with aaron's language set to Japanese)
- The issue has the Zulip Cloud footer linking to the help center article **Contact Support** and I just wanted to confirm that's the correct and we don't want to link directly to the support email address directly in the email.

<details>
<summary>Login notification email in Japanese</summary>

![Screenshot from 2023-09-26 15-31-02](https://github.com/zulip/zulip/assets/63245456/d4a105ce-9154-4dc3-b3af-9ab81a37ef2c)
</details>

---

**Screenshots and screen captures:**

<details>
<summary>Support deactivated realm - with Zulip Cloud footer</summary>

## Text version
![Screenshot from 2023-09-26 15-23-45](https://github.com/zulip/zulip/assets/63245456/ef504f30-df6a-43fe-80d0-1e664db1ee47)

## HTML version
![Screenshot from 2023-09-26 15-23-36](https://github.com/zulip/zulip/assets/63245456/8e5c19a4-737f-4027-b82f-cfd89a8c9cf5)
</details>
<details>
<summary>Active owner deactivated realm - with self-hosted footer</summary>

## Text version
*1st email is to the other active owner, 2nd email is to the owner who deactivated the organization.*
![Screenshot from 2023-09-26 14-39-33](https://github.com/zulip/zulip/assets/63245456/7e6ed8dc-e2ad-4c5d-8fce-cc410a86795d)

## HTML version
*1st email is to the other active owner, 2nd email is to the owner who deactivated the organization.*
![Screenshot from 2023-09-26 14-39-20](https://github.com/zulip/zulip/assets/63245456/717f1783-3f94-4104-9f6e-91e0e7464235)

</details>

---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented:
For localization of the date description, Python (like other languages) has libraries that will localized date-times; see https://lokalise.com/blog/date-time-localization/ for examples of how to do that. If we send a single email to all administrators, I guess that would need to be in the organization's default language; or we could have a loop that sends emails in each target user's individual language.

timabbott commented on zerver/actions/realm_settings.py:
```diff
@@ -552,3 +560,73 @@ def do_send_realm_reactivation_email(realm: Realm, *, acting_user: Optional[User
+            language=language,
+            context=not_owner_context,
+            realm=realm,
+        )
```
> This flag is optional for the deactivate_realm management command, but as there is no active user passed in that case, then the email is sent without a reference to who deactivated the realm.

If desired think we could say "The {EXTERNAL_HOST} administrators" as who deactivated this organization. I think in some sense for Zulip Cloud we may want this to say something like "Zulip support" or "Zulip account" as who did the deactivation if done via management command.

timabbott replied:
I guess we already have "Zulip account security" as the email sender.

laurynmm replied:
I didn't change the text of the emails for anonymous deactivation via support or management command. See update and updated screenshots below.

timabbott replied:
Works for now, we can always do that in a follow-up.

timabbott commented on zerver/actions/realm_settings.py:
```diff
@@ -552,3 +560,73 @@ def do_send_realm_reactivation_email(realm: Realm, *, acting_user: Optional[User
+                language=language,
+                context=other_owners_context,
+                realm=realm,
+            )
```
If we decide to turn this into a loop over owners, the logic might actually get simpler since we can do stuff like just set `is_acting_owner` in a context object constructed in each line.

You can probably also replace the `.copy` stuff with this pattern:
```
other_owners_context = dict(
    deactivating_owner=acting_user.full_name,
    ...
    **shared_context,
)
```

laurynmm replied:
Updates have us loop over the owners which is indeed simpler logic for setting the email context.

timabbott commented on templates/zerver/emails/realm_deactivated.html:
```diff
@@ -0,0 +1,26 @@
+</p>
+<p>
+    {% if corporate_enabled %}
+        {% trans contact_support="https://zulip.com/help/contact-support" %}If you have any questions or concerns, please <a href="{{ contact_support }}">contact Zulip Support</a> as soon as possible.{% endtrans %}
```
> The issue has the Zulip Cloud footer linking to the help center article Contact Support and I just wanted to confirm that's the correct and we don't want to link directly to the support email address directly in the email.

@alya thoughts on this detail?  `git grep help/contact-support templates/zerver/emails/` suggests we don't link to that page elsewhere in emails, and we probably do want a direct email reply in most cases ... so it could make sense to just rewrite both cases with "please reply to this email as soon as possible".

alya replied:
Yeah, that sounds good as long as the from: email will be correct for just sending a reply.

laurynmm replied:
I updated this so that the link in the Zulip Cloud email is a "mailto" link to the support email address.

If we instead want to be able to reply to these emails, then we'd set the from address to `FromAddress.SUPPORT`?

timabbott replied:
I think we can set `reply-to` of `FromAddress.SUPPORT` independently from `From`, if that's helpful.

laurynmm commented:
**Updates**:
- Loops over organization owners to set language and context variables for emails.
- Adds `localize` filter from `django.utils.formats` to the jinja envs filters so that the date is localized/internationalized for the owner's default language.
- Zulip Cloud emails have a link to the support email address and not the help article about contacting support.
- The anonymous deactivation emails have not been updated to have text about who deactivated the organization; see final set of screenshots below.

<details>
<summary>Strings added for translation</summary>

```
#: templates/zerver/emails/realm_deactivated.html:11
#: templates/zerver/emails/realm_deactivated.txt:3
"You have deactivated your Zulip organization, %(realm_name)s, on "
"%(localized_date)s."

#: templates/zerver/emails/realm_deactivated.html:13
#: templates/zerver/emails/realm_deactivated.txt:5
"Your Zulip organization, %(realm_name)s, was deactivated by "
"%(deactivating_owner)s on %(localized_date)s."

#: templates/zerver/emails/realm_deactivated.html:15
#: templates/zerver/emails/realm_deactivated.txt:7
"Your Zulip organization, %(realm_name)s, was deactivated on "
"%(localized_date)s."

#: templates/zerver/emails/realm_deactivated.html:20
"If you have any questions or concerns, please <a href=\"mailto:"
"%(support_email)s\">contact Zulip Support</a> as soon as possible."

#: templates/zerver/emails/realm_deactivated.html:22
#: templates/zerver/emails/realm_deactivated.txt:13
"If you have any questions or concerns, please contact your Zulip server "
"administrator at %(support_email)s as soon as possible."

#: templates/zerver/emails/realm_deactivated.subject.txt:1
"Your Zulip organization %(realm_name)s has been deactivated"

#: templates/zerver/emails/realm_deactivated.txt:11
"If you have any questions or concerns, please contact Zulip Support "
"(%(support_email)s) as soon as possible."

```
</details>

---

**Updated screenshots**:

<details>
<summary>Zulip Cloud deactivation emails with Spanish and Japanese owners</summary>

### Text version
![Screenshot from 2024-06-18 16-52-28](https://github.com/zulip/zulip/assets/63245456/25af8928-0a5a-405b-b0af-a6152c09f8a4)

### HTML version
![Screenshot from 2024-06-18 16-52-48](https://github.com/zulip/zulip/assets/63245456/6c9de583-d591-4ec8-9f83-2af75d55cff3)
![Screenshot from 2024-06-18 16-53-02](https://github.com/zulip/zulip/assets/63245456/befb26d9-2d33-4ed2-9865-5716f54c2c9d)
![Screenshot from 2024-06-18 16-53-19](https://github.com/zulip/zulip/assets/63245456/25669fa8-e3ae-48fa-92a4-64ab2e08c168)
</details>
<details>
<summary>Self-hosted deactivation emails with German and Russion owners</summary>

### Text version
![Screenshot from 2024-06-18 17-01-28](https://github.com/zulip/zulip/assets/63245456/dc1f2b54-696d-4b9a-9bde-21f94c3e7df7)

### HTML version
![Screenshot from 2024-06-18 17-01-48](https://github.com/zulip/zulip/assets/63245456/a18f57c6-2450-4f9f-ad87-4563d04eba7f)
![Screenshot from 2024-06-18 17-02-00](https://github.com/zulip/zulip/assets/63245456/2e265044-e6e9-44bb-b091-5f8846ae7f28)
![Screenshot from 2024-06-18 17-02-12](https://github.com/zulip/zulip/assets/63245456/a1be7d63-63c0-448f-b9c0-afab96b402fd)
</details>
<details>
<summary>Anonymous deactivation from support view - Zulip Cloud</summary>

### Text version
![Screenshot from 2024-06-18 17-17-51](https://github.com/zulip/zulip/assets/63245456/62cad411-9fd0-42d8-a313-93da2f177cfa)

### HTML version
![Screenshot from 2024-06-18 17-17-42](https://github.com/zulip/zulip/assets/63245456/d2aa0298-e33b-4585-8a20-ca81a06980d5)
</details>

alya commented:
The "contact Zulip support" links look weird to me in an email...

In other emails, I see:

> If you did not request this change, please contact us immediately at <{{ support_email }}>.

So how about this for the Zulip Cloud email here?

> If you have any questions or concerns, please contact us at <{{ support_email }}> as soon as possible.

laurynmm commented:
Updated Zulip Cloud contact us line for @alya's note above ...

<details>
<summary>HTML version, English, self-deactivated on Zulip Cloud</summary>

![Screenshot from 2024-06-18 18-20-45](https://github.com/zulip/zulip/assets/63245456/51510c6d-00df-48c6-9680-9df89bbb05fb)
</details>
<details>
<summary>Text version, English, self-deactivated on Zulip Cloud</summary>

![Screenshot from 2024-06-18 18-20-54](https://github.com/zulip/zulip/assets/63245456/702df989-8ff7-45b7-a442-0057e5cbdc71)
</summary>

alya commented:
Looks good to me, thanks!

laurynmm commented:
**Updates**:
- Changed the email `from_address` and `reply_to_email` to be the support email address for the server.
  - The `from_name` is still the security email. This is consistent with the "notify_change_in_email" from name and email.
- Updated the final line of the email to be the same for both Zulip Cloud and self-hosted servers. The translated string now reads: "If you have any questions or concerns, please reply to this email as soon as possible."

<details>
<summary>Updated screenshot of email in dev environment</summary>

![Screenshot from 2024-06-26 17-51-16](https://github.com/zulip/zulip/assets/63245456/60279e08-1a02-40c4-b8d2-c4f3e8e85521)
</details>

timabbott commented on zerver/actions/realm_settings.py:
```diff
@@ -800,3 +810,65 @@ def do_send_realm_reactivation_email(realm: Realm, *, acting_user: Optional[User
+            to_emails=[owner.delivery_email],
+            from_name=FromAddress.security_email_from_name(language=owner.default_language),
+            from_address=FromAddress.SUPPORT,
+            reply_to_email=FromAddress.SUPPORT,
```
I don't think there's ever a reason to set `reply_to_email` to be the same as `from_address` since it's basically just an override for `from_address`.

timabbott commented:
Dropped the `reply_to_address` as suggested in https://github.com/zulip/zulip/pull/26863#discussion_r1655347550 and marked to merge once CI passes again.
